[ARM][Clang] Make `+nosimd` functional for AArch32 Targets

Stylie777 opened the pull request:
`+simd` and `+nosimd` are used to enable or disable NEON Instructions
when compiling for AArch32 Targets. However, up until now, using these
has not been possible. To enable this, these options are mapped to the
relevant LLVM backend option (`+neon` and `-neon`) so it can be both
enabled and disabled successfully by the user.

Tests have been added to ensure this behaviour is maintained in the future,
along with updates to existing tests as behaviour has now changed relating
to the use of `+simd` and `+nosimd`.

As `simd` has been mapped within the ARMTargetParser.def, support for this
extension is also added for the `--print-support-extensions` command when
the target is AArch32. This will print the `simd` option, along with the
description that relates to the Neon feature. This previously was not
possible as `simd` did not have a related Feature or Negative Feature.

To make this functional as intended, MVE and MVE.FP now rely on their own
Enum identifier, rather than `AEK_SIMD`. While SIMD does refer to both
Neon and Helium technologies, in terms of command line options, SIMD relates
to Neon. Helium relates to MVE and MVE.FP. The Enum now reflects this too.

Stylie777 committed:
[Clang][ARM] Enable the use of `+simd` and `+nosimd` for AArch32 Targets

`+simd` and `+nosimd` are used to enable or disable NEON Instructions
when compiling for AArch32 Targets. However, up until now, using these
has not been possible. To enable this, these options are mapped to the
relevant LLVM backend option (`+neon` and `-neon`) so it can be both
enabled and disabled successfully by the user.

Tests have been added to ensure this behaviour is maintained in the future,
along with updates to existing tests as behaviour has now changed relating
to the use of `+simd` and `+nosimd`.

As `simd` has been mapped within the ARMTargetParser.def, support for this
extension is also added for the `--print-support-extensions` command when
the target is AArch32. This will print the `simd` option, along with the
description that relates to the Neon feature. This previously was not
possible as `simd` did not have a related Feature or Negative Feature.

Stylie777 committed:
[NFC][ARM] Split SIMD identifier away from MVE

Previously, the use of MVE or MVE.FP would be defined
by using the `ARM::AEK_SIMD` identifier. SIMD relates to
the Cortex-A and Cortex-R extension that enables NEON
instructions, which is called MVE for Cortex-M. To enable the
linking of `+simd` and `+nosimd` to `+neon` and `-neon`
when using clang, MVE and MVE.FP can now be defined
using a unique enum identifier, rather than using the existing
identifier for SIMD.

This was originally planned to be merged as part of #130296 but
the changes made `+nosimd` an invalid argument, which, while not
having any functionality, was allowed in previous versions of
LLVM. To avoid regressions being introduced, this has been
combined with the fix for `+nosimd` on AArch32.

jthackray commented:
typo in commit message: s/emable/enable/

jthackray commented on llvm/docs/ReleaseNotes.md:
```diff
@@ -85,6 +85,9 @@ Changes to the AMDGPU Backend
 Changes to the ARM Backend
 --------------------------
+* The `+nosimd` attribute is now fully supported. Previously, this had no effect when being used with
+AArch32 targets, however will now disable NEON instructions being generated. The `simd` is also now
```
```suggestion
AArch32 targets, however this will now disable NEON instructions being generated. The `simd` option is also now
```

DavidSpickett commented:
> used to emable or disable

enable

Stylie777 commented:
The description is fixed. Thanks @DavidSpickett and @jthackray (I will publish a commit to your improvements once I have a green CI so I know all is good).

Stylie777 committed:
Improve Release Notes

jthackray reviewed (approved):
Thanks for the fixes. LGTM now.

Stylie777 commented:
Thanks @jthackray. The CI Failure seems to be unrelated.

davemgreen commented on llvm/include/llvm/TargetParser/ARMTargetParser.def:
```diff
@@ -334,8 +334,8 @@ ARM_CPU_NAME("cortex-r7", ARMV7R, FK_VFPV3_D16_FP16, false,
              (ARM::AEK_MP | ARM::AEK_HWDIVARM))
-ARM_CPU_NAME("cortex-r52", ARMV8R, FK_NEON_FP_ARMV8, false, ARM::AEK_NONE)
-ARM_CPU_NAME("cortex-r52plus", ARMV8R, FK_NEON_FP_ARMV8, false, ARM::AEK_NONE)
+ARM_CPU_NAME("cortex-r52", ARMV8R, FK_NEON_FP_ARMV8, false, ARM::AEK_SIMD)
+ARM_CPU_NAME("cortex-r52plus", ARMV8R, FK_NEON_FP_ARMV8, false, ARM::AEK_SIMD)
```
Why are these two special-cased to include AEK_SIMD, unlike all the other cpus where it comes from FK_NEON_FP_ARMV8? Am I right in saying that removing it would not affect how the command line for -mcpu=cortex-r52 behaves? Or does it change the defaults in some way?

Stylie777 replied:
We support NEON on Cortex-R52 by default, rather than as an optional feature. When `+simd`/`+nosimd` was linked to `+neon` and `-neon`, I found that once the link between `simd` and `neon` was made, it was not passing `-target-feature +neon` as part of the cc1 command.

Let me investigate and see if we need this change or if it can determine `+neon` from the FPU, if not we may need to add `AEK_SIMD` to everything that should support it as it may not just be cortex-r52 that is affected here.

Stylie777 replied:
So, yes this is needed. Because `simd` is now linked to `neon`, if it is not explicitly defined for each Architecture/CPU that supports NEON, `-target-feature -neon` is passed by the compiler, disabling the feature. This is because when it parses the `.def` file, that CPU/Architecture is told SIMD is not supported as it is not included. I think it was a fluke it worked before, and it picked it up from the FPU in the backend, but the responsibility for determining if Neon is supported now lies with the front end with this change, so we need to explicitly define it. I will provide an update in due course that defines it for the Architecture's and CPU's that support NEON and tests to support it. I will also update the description of the PR to state this.

davemgreen replied:
I think I would have expected neon vs non-neon to come from the fpu option / default, under Arm. We (essentially) default to -mfpu=auto but it can get quite complex with all the different ways of specifying it.

Stylie777 replied:
I am posing this a question rather than a statement as I am not sure of the answer, but could LLVM be taking the target features defined as a `-target-feature` option over the features that come from the fpu? Reason I ask is before this change, `-target-feature +neon` never appeared for targets when I added `-###` to the command, but now they are. If neon is not defined for that target, it passes `-target-feature -neon` because the ID of `ARM::AEK_SIMD` is not included in the features from the `ARMTargetFeatures.def` file. 

That could explain why we need this change.

Stylie777 committed:
Move release notes to clang

Stylie777 reviewed (commented):
Thanks for the comments @davemgreen. I need to do some more investigation here as I think there is more going wrong than I initially thought after this change is introduced.

Stylie777 committed:
Fix release notes

Stylie777 committed:
Add ARM::AEK_SIMD for Arm V8-A+

To ensure NEON is activated for the ArmV8-A cores and
beyond, ARM::AEK_SIMD needs to be added for the
appropriate architectures. Appropriate tests have been
added to support this change.

Stylie777 commented:
I have just pushed an update for the Unit Tests that should turn the CI green. This is now ready for re-review @davemgreen

Stylie777 commented:
No worries 😀

If we remove `ARM::AEK_SIMD` is removed, the target will not be able to process or generate NEON instructions. I have added it to all ArmV8-a (and beyond) architectures and the Cortex-R52 CPU. This should cover all the Cortex-A and Cortex-R targets that support it. As ArmV7-A has NEON as an optional feature, I have not added it here. The only thing I may have missed is if Arm V7-A CPU's Support NEON as a mandatory feature, I would need to check this.

davemgreen commented:
NEON is never mandatory AFAIU in the architecture (FP too). We might assume it to be present though, as I believe it comes from the default -mfpu. (For example FK_CRYPTO_NEON_FP_ARMV8 from armv8-a).

Using something like this: https://godbolt.org/z/EKEMsaMdW. If I take this patch and remove the ARM::AEK_SIMD from the ARM_ARCH and ARM_CPU_NAME definitions it still seems to do OK. Is there some other reason that AEK_SIMD needs to be added to them? I tried some combinations of -march=..[+[no]fp] and -mfpu=... but they seemed OK and the tests passed. ARM::getFPUFeatures gets used in quite a few places though so something else might be going wrong.

Stylie777 commented:
Ok, thats interesting to know. To develop this, I have been using an assembly file with various AdvSIMD instructions, when ARM::AEK_SIMD is removed from the definitions I have found that it can no longer parse the instructions, giving an error that NEON is required. Once ARM::AEK_SIMD has been added to the definitions this keeps working.

It's entirely possible I have missed a change that needs making to the AsmParser, and this may be a better solution, but I found this change was required for the AsmParser to correctly determine that NEON was supported for that target.

Stylie777 committed:
Update LLVM Unit Tests

Stylie777 committed:
Ensure FPU Features are collected when using the Clang Assembler

Previously, FPU features were not collected when forming a list
of features for the Assembler. This also allows NEON to be pulled
from the FPU support rather than having it hardcoded into the
Target Parser Definitions.

davemgreen commented on clang/lib/Driver/ToolChains/Arch/ARM.cpp:
```diff
@@ -679,20 +679,18 @@ llvm::ARM::FPUKind arm::getARMTargetFeatures(const Driver &D,
         getARMSubArchVersionNumber(Triple) >= 7) {
       FPUKind = llvm::ARM::parseFPU("neon");
       (void)llvm::ARM::getFPUFeatures(FPUKind, Features);
+    } else if ((!Generic) ||
```
Can this now be `else if (!ForAS || !Generic) {`. I wasn't sure why armv7 is special? This should mean that `-march=armv8-m.main` for example works as it did before, and as far as I can tell the versions with cpus remain the same too. (i.e. clang/test/Driver/armv8.1m.main.s should no longer change).

davemgreen commented on clang/test/Driver/arm-fpu-selection.s:
```diff
@@ -0,0 +1,31 @@
+// used and tested for VFP and NEON Support
+
+// RUN: %clang -target arm-none-eabi -mcpu=cortex-r52 -c %s -o /dev/null | count 0
+// RUN: %clang -target arm-none-eabi -mcpu=cortex-r52 -c %s -o /dev/null -### 2> %t | FileCheck --check-prefix=CHECK-TARGET-FEATURES < %t %s
```
I think this might need to either pipe &2 into stdout, or `cat %t | FileCheck..`

davemgreen commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -38,6 +38,9 @@ Potentially Breaking Changes
 - Fix missing diagnostics for uses of declarations when performing typename access,
   such as when performing member access on a '[[deprecated]]' type alias.
   (#GH58547)
+- For ARM targets, when using cc1as, the features included in the selected CPU or
+  Arch's FPU are now loaded and utilized. If you wish not to use a specific feature,
+  this will need appending to the command line used.
```
What, from a user facing point of view, changes with the new patch? My understanding is that we pass the fpu features from the driver->cc1as now as opposed to calulcating them in cc1as, but it doesn't alter the behaviour otherwise? The examples I tried all work as before, but there are a lot of different possibilities so I might be missing one again.

davemgreen reviewed (commented):
Thanks - this looks good.

Stylie777 reviewed (commented):
Thanks @davemgreen for the re-review. I have made some updated based on your comments.

mstorsjo commented on clang/test/Driver/arm-fpu-selection.s:
```diff
@@ -0,0 +1,31 @@
+// activation of NEON for supported targets. The Cortex-R52 will be
+// used and tested for VFP and NEON Support
+
+// RUN: %clang -target arm-none-eabi -mcpu=cortex-r52 -c %s -o /dev/null | count 0
```
This requires a `// REQUIRES: arm-registered-target` (like `armv8.1.m.main.s` has), as this actually does try to compile the code.

Actually compiling code in Clang driver tests would ideally not be done (we should ideally only inspect the generated command lines), but we do seem to have some precedent for it already in `armv8.1.m.main.s`, and especially for the implicit behaviours that aren't necessarily visible in the command line, I guess there's no other good alternative.

So I guess this kind of test is fine, but add the `REQUIRES` line.

Stylie777 replied:
Done

Stylie777 commented:
I have rebased this now after #134612 has been merged. This PR is only now looking at enabling +nosimd for ARM targets.

davemgreen commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -419,6 +419,9 @@ Bug Fixes to Attribute Support
 - No longer crashing on ``__attribute__((align_value(N)))`` during template
   instantiation when the function parameter type is not a pointer or reference.
   (#GH26612)
+- The ``+nosimd`` attribute is now fully supported for AArch32. Previously, this had no effect when being used with
```
Can a comment be added to the arm section too / instead?

Stylie777 replied:
I have moved it to the ARM section.

davemgreen reviewed (approved):
Thanks. LGTM

Stylie777 committed:
Move release note to ARM section.

Stylie777 reviewed (commented):
Thanks @davemgreen for the review.
